dataservice/api/diagnosis/models.py

```diff
@@ -66,3 +69,65 @@ class Diagnosis(db.Model, Base):
                                db.ForeignKey('participant.kf_id'),
                                doc='the participant who was diagnosed',
                                nullable=False)
+    biospecimen_id = db.Column(KfId(),
+                               db.ForeignKey('biospecimen.kf_id'),
+                               doc='the biospecimen which recieved the '
+                               'diagnosis',
+                               nullable=True)
+
+
+def validate_diagnosis_biospecimen(target):
+    """
+    Ensure that both the diagnosis and biospecimen
+    (referred to by biospecimen_id) have the same participant
+
+    If this is not the case then raise DatabaseValidationError
+    """
+    from dataservice.api.biospecimen.models import Biospecimen
+    from dataservice.api.errors import DatabaseValidationError
+
+    # Return if diagnosis is None
+    if not target:
+        return
+
+    # Get biospecimen by id
+    bsp = None
+    if target.biospecimen_id:
+        bsp = Biospecimen.query.get(target.biospecimen_id)
+
+    # If biospecimen doesn't exist, return and
+    # let ORM handle non-existent foreign key
+    if bsp is None:
+        return
+
+    # Check if this diagnosis and biospecimen refer to same participant
+    if target.participant_id != bsp.participant_id:
+        operation = 'modify'
+        target_entity = Diagnosis.__tablename__
+        kf_id = target.kf_id or ''
+        message = (
+            ('a diagnosis cannot be linked with a biospecimen if they '
+             'refer to different participants. diagnosis {} '
+             'refers to participant {} and '
+             'biospecimen {} refers to participant {}')
+            .format(kf_id,
+                    target.participant_id,
+                    bsp.kf_id,
+                    bsp.participant_id))
+        raise DatabaseValidationError(target_entity, operation, message)
+
+
+@event.listens_for(Diagnosis, 'before_insert')
```

Comment:
Use ORM object events instead of object attribute events (i.e. set or validate). Saw weird and unpredictable behavior with attribute events during deserialization of JSON to ORM objects. 

Probably better to use `before_insert`, `before_update` events anyway because we only care about validating biospecimen_id before the commit happens. The frequency of occurrence on the attribute events is much greater and we don't need to check the value of biospecimen_id every time its changed.